### ISSUE TITLE
Unable To Reproduce Results in the Paper

### ISSUE THREAD

**Issue by yangky11** [ischlag/TP-Transformer#3]
Hi,

Thanks for releasing the code.
I ran the code for all of the `arithmetic` modules and was unable to get the testing performance reported in the paper (Figure 2).  For example, below is the training log for the `arithmetic__add_sub_multiple` module. The final testing accuracy is about 0.01, while the accuracy is very high in Figure 2. I was running `python main.py --model_name tp-transformer --module_name arithmetic__add_sub_multiple`. Maybe there is something wrong with the hyperparameters I was using? 

```
Taking scripts from /n/fs/pvl-mathqa/deepmind-mathematics/TP-Transformer
... and saving them in log/arithmetic__add_sub_multiple/tp-transformer/lr=0.0001_bs=256_h=512_f=2048_nl=6_nh=8_d=0.0_Adam_/195932581/source
Arguments:
	seed=195932581
	model_name='tp-transformer'
	module_name='arithmetic__add_sub_multiple'
	load_model=''
	eval_mode=False
	n_steps=10000
	max_strikes=1000
	log_every=50
	eval_every=200
	full_loader=False
	force_remove=False
	force_reload=False
	no_train=False
	log_folder='log/arithmetic__add_sub_multiple/tp-transformer/lr=0.0001_bs=256_h=512_f=2048_nl=6_nh=8_d=0.0_Adam_/195932581'
	log_suffix=''
	optimizer='Adam'
	learning_rate=0.0001
	beta1=0.9
	beta2=0.995
	batch_size=256
	max_abs_grad_norm=0.1
	grad_accum_steps=1
	dropout=0.0
	hidden=512
	n_layers=6
	n_heads=8
	filter=2048
	d_r=0
	device=device(type='cuda')
using JIT loader
Loading JIT datasets ...
Creating iterators ...
Loading JIT datasets ...
Creating iterators ...
Building model ...
done. 49178624 trainable parameters.

eval: loss=6.1017 acc=0.0000 b/sec=0.16 time=0.1 min
run:  log/arithmetic__add_sub_multiple/tp-transformer/lr=0.0001_bs=256_h=512_f=2048_nl=6_nh=8_d=0.0_Adam_/195932581
best: loss=6.1017 acc=0.0000 best since 0 steps and 0.0 min

training for 10000 steps ...
  50: loss=1.3392 acc=0.0706 steps/sec=2.34
 100: loss=1.1730 acc=0.0740 steps/sec=2.49
 150: loss=1.1631 acc=0.0755 steps/sec=2.51
 200: loss=1.1598 acc=0.0728 steps/sec=2.50

eval: loss=3.8948 acc=0.0094 b/sec=0.15 time=1.5 min
run:  log/arithmetic__add_sub_multiple/tp-transformer/lr=0.0001_bs=256_h=512_f=2048_nl=6_nh=8_d=0.0_Adam_/195932581
best: loss=3.8948 acc=0.0094 best since 0 steps and 0.5 min

 250: loss=1.1530 acc=0.0729 steps/sec=0.94
 300: loss=1.1450 acc=0.0758 steps/sec=2.56
 350: loss=1.1446 acc=0.0745 steps/sec=2.44
 400: loss=1.1416 acc=0.0771 steps/sec=2.54

eval: loss=4.2658 acc=0.0116 b/sec=0.16 time=3.4 min
run:  log/arithmetic__add_sub_multiple/tp-transformer/lr=0.0001_bs=256_h=512_f=2048_nl=6_nh=8_d=0.0_Adam_/195932581
best: loss=3.8948 acc=0.0094 best since 200 steps and 1.9 min

 450: loss=1.1411 acc=0.0737 steps/sec=1.83
 500: loss=1.1396 acc=0.0781 steps/sec=2.51
 550: loss=1.1378 acc=0.0734 steps/sec=2.49
 600: loss=1.1372 acc=0.0767 steps/sec=2.53

eval: loss=4.3701 acc=0.0120 b/sec=0.15 time=4.8 min
run:  log/arithmetic__add_sub_multiple/tp-transformer/lr=0.0001_bs=256_h=512_f=2048_nl=6_nh=8_d=0.0_Adam_/195932581
best: loss=3.8948 acc=0.0094 best since 400 steps and 3.3 min

 650: loss=1.1337 acc=0.0725 steps/sec=1.92
 700: loss=1.1340 acc=0.0753 steps/sec=2.46
 750: loss=1.1346 acc=0.0789 steps/sec=2.53
 800: loss=1.1344 acc=0.0780 steps/sec=2.51

eval: loss=4.3407 acc=0.0112 b/sec=0.15 time=6.2 min
run:  log/arithmetic__add_sub_multiple/tp-transformer/lr=0.0001_bs=256_h=512_f=2048_nl=6_nh=8_d=0.0_Adam_/195932581
best: loss=3.8948 acc=0.0094 best since 600 steps and 4.8 min

 850: loss=1.1386 acc=0.0759 steps/sec=1.93
 900: loss=1.1298 acc=0.0775 steps/sec=2.51
 950: loss=1.1329 acc=0.0785 steps/sec=2.51
1000: loss=1.1335 acc=0.0754 steps/sec=2.51

eval: loss=4.2900 acc=0.0119 b/sec=0.15 time=7.7 min
run:  log/arithmetic__add_sub_multiple/tp-transformer/lr=0.0001_bs=256_h=512_f=2048_nl=6_nh=8_d=0.0_Adam_/195932581
best: loss=3.8948 acc=0.0094 best since 800 steps and 6.2 min

1050: loss=1.1347 acc=0.0786 steps/sec=1.92
1100: loss=1.1278 acc=0.0791 steps/sec=2.51
1150: loss=1.1296 acc=0.0782 steps/sec=2.53
1200: loss=1.1326 acc=0.0767 steps/sec=2.52

eval: loss=4.9544 acc=0.0104 b/sec=0.15 time=9.1 min
run:  log/arithmetic__add_sub_multiple/tp-transformer/lr=0.0001_bs=256_h=512_f=2048_nl=6_nh=8_d=0.0_Adam_/195932581
best: loss=3.8948 acc=0.0094 best since 1000 steps and 7.6 min

1000 steps without improvement. Exiting.
```

**Comment by ischlag** [comment]
Hi yangky11, 

Note the train.sh file with the hyper-parameters used during training. Especially the --full_loader argument is important. **We have done all experiments with the full_loader argument.** The dynamic loader was added at a later stage for convenience but since then we noticed an issue with it, i.e. _the dynamic loader seems to be not working quite as expected_. Unfortunately, I'm currently busy with other projects and don't have time to debug the dynamic loader. For now, please just use the full loader as we did.  Same for arithmetic__add_or_sub. 

If you could use the command in the train.sh file and change the module name to the one of your interest then training should work just fine. If I recall correctly, the module "numbers__place_value" is fairly simple should achieve 100% accuracy very quickly. 

And finally, I'd like to point out that we trained on all modules simultaneously. This can be achieved if you train on the module "all_modules". However, this will require a lot of RAM memory to load the dataset at the beginning (this took about 20min). If you are only interested in individual tasks or a subset of tasks, then you can likely profit from a smaller model. For this, I'd recommend trying a smaller h and/or fewer layers. 

Imanol

**Comment by yangky11** [comment]
I'll try it out. Thanks!

Also, should I use the same hyperparameters for the transformer baseline (not tp-transformer)?

**Comment by ischlag** [comment]
Yes, absolutely.
